File: backend/src/main/java/dev/zeroday/health/habits/model/Habit.java
package dev.zeroday.health.habits.model;

import dev.zeroday.health.common.BaseEntity;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.LocalTime;
import java.util.List;

@Entity
@Table(name = "habit")
@Getter
@Setter
@NoArgsConstructor
public class Habit extends BaseEntity {

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(nullable = false)
    private String name;

    private String description;

    @Column(nullable = false, length = 50)
    private String frequency;

    @Column(name = "target_count")
    private Integer targetCount;

    @JdbcTypeCode(SqlTypes.ARRAY)
    @Column(name = "days_of_week", columnDefinition = "integer[]")
    private List<Integer> daysOfWeek;

    @Column(length = 20)
    private String color;

    @Column(length = 50)
    private String icon;

    @Column(nullable = false)
    private boolean active = true;

    @Column(name = "habit_type", nullable = false, length = 10)
    private String habitType = "GOOD";

    @Column(name = "target_days", nullable = false)
    private int targetDays = 66;

    @Column(length = 50)
    private String category;

    @Column(columnDefinition = "text")
    private String cue;

    @Column(columnDefinition = "text")
    private String routine;

    @Column(columnDefinition = "text")
    private String reward;

    @Column(name = "stack_after_habit_id")
    private Long stackAfterHabitId;

    @Column(length = 20)
    private String difficulty = "MEDIUM";

    private Integer priority;

    @Column(name = "reminder_time")
    private LocalTime reminderTime;
}

File: frontend/src/types/habit.ts
export type HabitType = 'GOOD' | 'BAD';
export type HabitDifficulty = 'EASY' | 'MEDIUM' | 'HARD';

export const HABIT_CATEGORIES = [
  'Health', 'Fitness', 'Productivity', 'Mindfulness',
  'Social', 'Financial', 'Learning', 'Custom',
] as const;

export const MILESTONE_THRESHOLDS = [7, 14, 21, 30, 60, 66, 90, 100, 180, 365] as const;

export interface Habit {
  id: number;
  name: string;
  description: string | null;
  frequency: string;
  targetCount: number;
  daysOfWeek: number[];
  color: string | null;
  icon: string | null;
  active: boolean;
  habitType: HabitType;
  targetDays: number;
  currentStreak: number;
  longestStreak: number;
  completionRate: number;
  daysSinceLastOccurrence: number;
  totalOccurrences: number;
  formationProgress: number;
  category: string | null;
  cue: string | null;
  routine: string | null;
  reward: string | null;
  stackAfterHabitId: number | null;
  difficulty: string | null;
  priority: number | null;
  reminderTime: string | null;
}

export interface HabitLog {
  date: string;
  completed: boolean;
  notes: string | null;
  intensity: number | null;
  mood: string | null;
  context: string | null;
  skipReason: string | null;
}

export interface HabitMilestone {
  id: number;
  habitId: number;
  milestoneType: string;
  milestoneValue: number;
  achievedAt: string;
}

export interface DailyHabitStatus {
  habitId: number;
  habitName: string;
  habitType: HabitType;
  completed: boolean;
  currentStreak: number;
  daysSinceLastOccurrence: number;
  formationProgress: number;
  targetDays: number;
  color: string | null;
}
